### PR TITLE
OSPP Implementation: Efficient Federated Fine-Tuning Framework for Large Language Models on KubeEdge-Ianvs 

### LINKED ISSUE #198: Large Language Model Federated Fine-tuning Algorithm Based on KubeEdge-Ianvs
**Description**:
With the increasing adoption of Large Language Models (LLMs) in privacy-sensitive domains such as healthcare, finance, and government services, fine-tuning LLMs with local data to meet domain-specific requirements has become essential. However, traditional federated learning (FL) methods struggle to scale due to the massive computational and communication costs of LLMs. This project aims to build an efficient federated fine-tuning framework for LLMs on KubeEdge-Ianvs, supporting parameter-efficient fine-tuning (PEFT) algorithms such as LoRA and Adapters. Leveraging KubeEdge’s capabilities in edge node orchestration and resource management, the system will enable low-cost, privacy-preserving LLM fine-tuning across organizations.

**Expected Outcomes**:
The goal of this work is to design and implement a new federated learning paradigm within the KubeEdge-Ianvs framework, drawing on the experience of KubeEdge-Sedna. By establishing this paradigm, the project aims to support scalable and efficient training of large models in edge computing scenarios. To this end, a baseline federated fine-tuning algorithm for large models will be developed, alongside a standardized testing suite to ensure reproducibility and comprehensive evaluation.
1. Implement a new federated learning process paradigm in KubeEdge-Ianvs, referencing the federated learning algorithm library from KubeEdge-Sedna.
2. Based on the new federated learning paradigm in KubeEdge-Ianvs, implement a large-model federated fine-tuning algorithm as a baseline.
3. Based on the new federated learning paradigm in KubeEdge-Ianvs, provide a standardized testing suite, including datasets, evaluation metrics, and testing environment scripts.
4. (Optional) Optimize the large-model federated fine-tuning algorithm baseline in the paradigm for local training on edge clients.

**Recommended Skills**:
KubeEdge-Ianvs, Python, LLMs

**Useful links**:
[Introduction to lanvs](https://ianvs.readthedocs.io/en/latest/distributed-synergy-ai-benchmarking.html)
[Quick Start](https://ianvs.readthedocs.io/en/latest/guides/quick-start.html)
[How to test algorithms with Ianvs](https://ianvs.readthedocs.io/en/latest/guides/how-to-test-algorithms.html)
[Example LLMs Benchmark List](https://github.com/terryyz/llm-benchmark)
[Ianvs v0.1 documentation](https://ianvs.readthedocs.io/en/latest/)
[Example Federated LLM](https://github.com/FederatedAI/FATE-LLM)

### PR DESCRIPTION
**Pull request by wwhwhu**
https://github.com/user-attachments/assets/88449ba1-65c0-4c86-b1c4-6b64b1600b07

**What type of PR is this?**
/kind feature
/kind design
/kind api-change

---
**What this PR does / why we need it**:
This PR introduces a new federated LLM fine-tuning algorithm (`federated_llm_learning`) into Ianvs, extending the existing `federated_learning` paradigm. The motivation is to support parameter-efficient fine-tuning (PEFT) of large language models (LLMs) under federated learning, which is not currently available in Ianvs.
Key features:

- **New algorithmic mode (`if_mode_llm`)**  
  Switches between standard federated learning and federated LLM fine-tuning seamlessly.  

- **Integration of PEFT methods**  
  - LoRA adapters  
  - P-Tuning v2 prompt embeddings  

- **Aggregation strategies**  
  - FedAvg for LoRA/P-Tuning modules  
  - FedAvgM (with server momentum) for stable convergence  

- **GPU-aware task scheduler**  
  - Queue-based multi-GPU scheduling  
  - Each GPU handles at most one client at a time  
  - Prevents OOM while scaling to dozens of clients on a single-node multi-GPU server  

- **Benchmarking metrics**  
  - ROUGE-1, ROUGE-2, ROUGE-L  
  - BLEU-4  
 
---
**Which issue(s) this PR fixes**:  
Fixes #198
Also fixes minor typos in `sedna_federated_learning.py` (no functional change).

---

**Special notes for your reviewer**:

- Backward compatibility is preserved:  
  All PEFT-specific logic is sandboxed in the new algorithm and `examples/federated-llm/*`. New `federated_llm_learning` algorithm (with `if_mode_llm` flag). Existing FL paradigms remain untouched.
- Dataset utilities extended:  
  Added JSONL support and `partition_llm_data()` for federated dataset splits.
- Example directory includes ready-to-run jobs:  
  `examples/federated-llm/fedllm-peft` contains YAML configs, algorithm code, and metric scripts.

### CODE REVIEW

File: examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        clients: 
+            All clients in federated learning job
+
+        Returns
+        -------
+        update_weights : Array-like
+            final weights use to update model layer
+        """
+
+        print("begin aggregation....")
```
Comment (hsj576) [inline examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py]: LOGGER.info should be used instead of print.

File: examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        for k, v in clients[0].weights.items():
+            v0 = torch.as_tensor(v, dtype=v.dtype if isinstance(v, np.ndarray) else v.dtype,
+                                    device=v.device if torch.is_tensor(v) else "cpu")
+            agg = torch.zeros_like(v0, dtype=torch.float32)
+            for c in clients:
+                vc = torch.as_tensor(c.weights[k], dtype=agg.dtype, device=agg.device)
+                agg += vc * (c.num_samples / self.total_size)
+            aggregated[k] = agg
+        self.weights = deepcopy(aggregated)
+        print("finish aggregation....")
```
Comment (hsj576) [inline examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py]: LOGGER.info should be used instead of print.

File: examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py (hunk 1 above)
Comment (wwhwhu) [reply examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py]: Thanks for the suggestion! Updated the code to use LOGGER.info() instead of print().

File: examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py (hunk 2 above)
Comment (wwhwhu) [reply examples/federated-llm/fedllm-peft/algorithm/FedAvg-PEFT.py]: Thanks for the suggestion! Updated the code to use LOGGER.info() instead of print().

**Review by hsj576 — APPROVED** [review]
/lgtm
